### DISCUSSION TITLE
nodal - FV variable coupling

### DISCUSSION THREAD

**Discussion by federicohattab** [idaholab/moose#23494] [Q&A Modules: Navier-Stokes] [answered]
Hello,

I am a new user.
I am trying to model the diffusion of a specie through a pipe and into the fliud inside the pipe.
The geometry is a 2D rectangle divided into 2 subodmains, representing half axial section of the pipe (wall + fluid domain). The fluid domain is modeled with the NavierStokesFV module. For the solid domain I am using the diffusion, time derivative and some nodal kernels to take into account trapping and releasing of the specie. Inside the NavierStokesFV module the specie is represented by the scalar equation.

I am having difficulties in defining the coupling of the specie concentration at the interface between the fluid and solid domain, because on one side I have a nodal variable while on the other side a FV variable. 

Any tip or general direction on how to achieve the coupling?
Thank you

**Comment by GiudGiud** [comment, marked as answer]
Hello

For now we cant couple in the same mesh finite volume and finite element. 

This is on our tasks for this year, getting both FE-FV coupling in the system and the option to have two numerical systems solved together on the same mesh. There's a lot of progress on the second task, you could already attempt to do it that way (and do fixed point iterations between the two systems). 

If you're not keen on developing, you should consider using a multiapp approach. You can have two applications, duplicate the mesh, and transfer the fields from one app to the other app in an auxiliary field of the compatible finite element/volume type. There will be some projection error but you can use a renormalization to be globally conservative. 

See this tutorial for more on multiapps & transfers
https://mooseframework.inl.gov/getting_started/examples_and_tutorials/tutorial02_multiapps/presentation/index.html#/
https://mooseframework.inl.gov/syntax/MultiApps/index.html
Guillaume

**Reply by federicohattab** [reply]
Thank you for the reply. Would it be better to do also the fluid domain with FE and simulate the specie with a time derivative, diffusion and advection kernel? (I am not sure this is possible since there is a FV scalar advection kernel but not a FE one)

**Reply by GiudGiud** [reply]
You can do that.
We have not added the FE scalar advection kernel but they are very similar to the energy advection kernel if you want to try

**Reply by federicohattab** [reply]
thank you, the implementation of an advection kernel is straight forward indeed. I am using the INSADEnergyAdvection as a reference. It works fine if I input manually a velocity vector in the input (like in the tutorial Example2), however I am having difficulties in retrieving the fluid velocity vector inside the Kernel. What is the method to get the INS fluid velocity vector?

**Reply by GiudGiud** [reply]
INSAD is using material properties for residuals. You ll want to have a look at adding a residual that way too, in order to be able to do the SUPG stabilization

**Reply by federicohattab** [reply]
Thanks, seems to work great, however I am now struggling in obtaining a conservation balance for the scalar in the fluid field. I am sorry if the question is obvious, but how would you verify the conservation of a scalar at the boundary of a 2D rectangular domain?

Right now, I am using a simplified case where the fluid is still (u = 0, v = 0) and there is a DirichletBC on the left, right and top boundaries equal to 10, 1 and 1 respectively (for the scalar). On the bottom boundary there is a NeumannBC for the scalar (i.e. a wall). I can't manage to obtain a net scalar flux across the boudaries equal to zero (the volume reached saturation since the overall concentration in the domain is constant).

For computing the diffusive fluxes at the interface I am usign the SideDiffusiveFluxIntegral. It is not clear to me what is the best option to get the advective flux (possibly the "VolumetricFlowRate" postprocessor, specifying as advected variable my scalar).

Thank you for the help!

**Reply by GiudGiud** [reply]
You can use the VolumetricFlowRate postprocessors on boundaries around the domain to evaluation the conservation. It will let you specify which quantity is being advected and what the advecting velocity is. 

However, this is looking at surface balance and the FE method is conservative over volumes. 
To check that, we implemented volumetric conservation postprocessors, such as this one
https://mooseframework.inl.gov/source/postprocessors/INSElementIntegralEnergyAdvection.html
You should be able to adapt it for scalar quantities. 

The SideDiffusiveFluxIntegral is not going to be valid for all types of diffusive fluxes. You ll need to double check the definition of the diffusive flux in your kernel and in the postprocessor.

**Reply by federicohattab** [reply]
Thank you. I adapted the linked postprocessor. So it is normal that the outlet VolumetricFlowRate is off by a bit (e.g. 0.5 - 1%) compared to the inlet VolumetricFlowRate, even though the volumetric conservation postprocessor returns a basically zero (1e-17) value.

To my understanding the SideDiffusiveFluxIntegral should be valid for the "ADHeatConduction" kernel (correct me if wrong please), but I am not sure about the "CoefDiffusion" one

**Reply by GiudGiud** [reply]
Yes it s normal. Does it get better with refinement?

**Reply by federicohattab** [reply]
I verified that the quantity of scalar entering the domain due to the DirichletBC heavily depends on the resolution of the mesh (it has a variation of about 20%).
Using adaptive mesh doesn't change this, but causes the outlet VolumetricFlowRate to be very similar to the inlet one (given the indicator I used shown below, this is less true for longer simulations, as the mesh tends to be coarser due to the system being stationary).

1. To have a proper indication of the quantities entering/leaving my domain I need to develop a volumetric conservation postprocessor only for the volumes at the boundaries I guess (I didn't find a similar postprocessor in the documentation).
2. For verifying the diffusion term, is it ok to used the SideDiffusiveFluxIntegral if my diffusion kernel is a modified version of "ADHeatConduction" containing only the quantity being diffused and the diffusivity?

Thank you

```
[Adaptivity]
   max_h_level = 3
   marker = marker
   [Indicators]
     [indicator]
       type = GradientJumpIndicator
       variable = myscalar
     []
   []
   [Markers]
     [marker]
       type = ErrorFractionMarker
       indicator = indicator
       refine = 0.2
       coarsen = 0.2
     []
   []
 []
```

**Reply by GiudGiud** [reply]
You may want to make the refinement permanent instead of using adaptivity. To do that you can use a RefineBlockGenerator in targetted blocks.

If you care about conserving an inlet flux, you should consider a Neumann boundary condition. 

1) A volumetric conservation postprocessor will not show you the quantities entering or leaving a block. It will just compute the advection term contribution (with or without diffusion, depending on how you write it) to the balance, which can tell you about the importance of other terms (if you have a reaction term for example)

2) What kind of boundary are you using the SideDiffusiveFlux on ? Dirichlet ? No BC?

**Reply by federicohattab** [reply]
![image](https://user-images.githubusercontent.com/124260728/223391539-8d094204-0ebc-4625-b9dc-aa690d94469d.png)

- These are the boundary conditions I am using. The fluid flow is upwards in the vertical direction, while C enters from the left.
- On the bottom I am using an ImplicitNeumannBC because if I use a NeumannBC at the first iteration I get C = 5 in the entire fluid domain, which is not physical, I don't understand why.
- Even though I impose no convective flux and no gradient at the bottom, I have a **non-negligible** diffusion flux and a convection flux passing thorugh the bottom boundary... this is my biggest corcern right now.
- I am using the SideDiffusiveFlux on all four boundaries. the fluid domain doesn't reach saturation due to the low diffusivity of the scalar and the velocitiy of the fluid, and as expected the diffusion fluxes on the top and right boundaries are very low.

Thank your for your time!

**Reply by GiudGiud** [reply]
1) C entering with a Dirichlet BC and 0 velocity seems odd. Is this going to be through a purely diffusive process?

2) With a Neumann BC of 0 you get an influx of C at the bottom?

Of note here, the C=100 BC is applied on the bottom left corner, and there's also a v=0.1m/s velocity there, so there is flux from that

3) With two BCs essentially achieving the same 0 flux, I'd be concerned this is not valid
To check all this I would:
- refine the bottom, see if the situation improves
- add an empty region at the bottom that has the velocity but does not connect with the C=100 boundary
Then we could see if the C concentration is entering the domain through the bottom

Are you still using finite volume?
